python/cuspatial/cuspatial/geometry/geoseries.py

```diff
@@ -0,0 +1,792 @@
+        """
+        The GeoArrow column format for GpuPolygons uses the same scheme as the
+        format for LineStrings - MultiPolygons and Polygons from the same
+        GeoSeries (or another data source) are stored in the same contiguous
+        buffer. Rings are stored in the offsets array, exterior/interior
+        polygons are stored in the `polys` array in shapefile format, and the
+        `mpolys` array of pairs determines which polys are members of
+        MultiPolygons.
+        """
+        # GpuPolygonArray uses the offsets buffer for rings!
+        super().__init__(xy, rings, z)
+        self._polys = polys
+        self._mpolys = mpolys
+
+    @property
+    def rings(self):
+        """
+        Each polygon in the GeoSeries is specified in the rings buffer, which
+        follows the same format as `offsets` in GpuOffsetArray.
+        """
+        return self.offsets
+
+    @rings.setter
+    def rings(self, rings):
+        self.offsets = rings
+
+    @property
+    def polys(self):
+        """
+        Polygons are specified according to the shapefile format. The first
+        ring in any polygon is its exterior ring, all subsequent rings are
+        interior rings. The `polys` column specifies which rings comprise each
+        polygon in the GeoSeries.
+        """
+        return self._polys
+
+    @polys.setter
+    def polys(self, polys):
+        self._polys = polys
+
+    @property
+    def mpolys(self):
+        """
+        Polygons that fall within bounds constrained by each pair of values in
+        this column are members of a MultiPolygon. Each pair of values
+        specifies a single MultiPolygon. All MultiPolygons in the GeoSeries
+        are contained here.
+        """
+        return self._mpolys
+
+    @mpolys.setter
+    def mpolys(self, mpolys):
+        self._mpolys = mpolys
+
+    def __repr__(self):
+        result = (
+            f"{super().__repr__()}"
+            f"polys:\n"
+            f"{self.polys}\n"
+            f"rings:\n"
+            f"{self.rings}\n"
+            f"mpolys:\n"
+            f"{self.mpolys}\n"
+        )
+        return result
+
+    def copy(self, deep=True):
+        """
+        See GpuCoordinateArray
+        """
+        result = super().copy(deep)
+        result.polys = self.polys.copy(deep)
+        result.rings = self.rings.copy(deep)
+        result.mpolys = self.mpolys.copy(deep)
+        return result
+
+
+class gpuGeometry:
+    def __init__(self, source, index):
+        """
+        The base class of individual GPU geometries. This and its inheriting
+        classes do not manage any GPU data directly - each gpuGeometry simply
+        stores a reference to the GeoSeries it is stored within and the index
+        of the geometry within the GeoSeries. Child gpuGeometry classes
+        contain the logic necessary to serialize and convert GPU data back to
+        Shapely. """
+        self._source = source
+        self._index = index
```

Comment:
I'm not sure about the design of this Class and the child classes. Storing an index into an underlying GeoSeries feels clumsy.

What does the actual cuDF object(s) underneath a cuSpatial GeoSeries look like?

Reply:
A `Series` of Points, a `Series` of MultiPoints with a `Series` of their offsets, a `Series` of LineString coordinates with a `Series` of their offsets and another `Series` for their MultiLine boundaries, and the same structure for Polygons plus one more `Series` for rings. A `GeoColumn` stores all of these coordinates. This is the GeoArrow format we've worked up with @trxcllnt and the geopandas crew - just GeoArrow `List<List>` types IIRC.

The existing cuspatial algorithms don't deal with individual geometries for anything, everything is a dense array of coordinates and geometry offsets, which is reflected here. When a user asks for a specific geometry, all that is returned is as you see the index and offsets used to compute that particular geometry, and the logic necessary to convert it back to GeoPandas.